### PR TITLE
Allow adding arbitrary files

### PR DESCRIPTION
**Pull request by mccheah**
Similarly to how jars can be uploaded from the client to the driver, arbitrary files should also be allowed to be mounted. This is similar to how Spark supports sending files via `spark.files` as well as jars.

Note that the files are added to the working directory of the driver because it would be necessary for the user to find these files in their application, and this matches the semantics of `spark.files`. However, if we de-duplicate the files with identical names, it would be difficult for the user to find the resolved files. Thus unlike the case with jars, we explicitly throw an error if the user gives two files with the same name.

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala [hunk 1]
```diff
@@ -94,7 +94,17 @@ package object config {
...
+        """.stripMargin)
+      .stringConf
+      .createOptional
+
+  private[spark] val KUBERNETES_DRIVER_UPLOAD_FILES =
+    ConfigBuilder("spark.kubernetes.driver.uploads.files")
+      .doc("""
+          | Comma-separated list of files to send to the driver and
           | all executors when submitting the application in cluster
           | mode.
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala]: note that these are placed in the CWD of the driver and executors once running

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/CompressionUtils.scala [hunk 2]
```diff
@@ -46,7 +47,8 @@ private[spark] object CompressionUtils extends Logging {
    * @param paths A list of file paths to be archived
    * @return An in-memory representation of the compressed data.
    */
-  def createTarGzip(paths: Iterable[String]): TarGzippedData = {
+  def createTarGzip(paths: Iterable[String]):
+      TarGzippedData = {
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/CompressionUtils.scala]: format-only change -- is this for style?

**Comment by mccheah** [comment]
Forgot documentation - will add that shortly

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala [hunk 3]
```diff
@@ -94,7 +94,17 @@ package object config {
   private[spark] val KUBERNETES_DRIVER_UPLOAD_JARS =
     ConfigBuilder("spark.kubernetes.driver.uploads.jars")
       .doc("""
-          | Comma-separated list of jars to sent to the driver and
+          | Comma-separated list of jars to send to the driver and
+          | all executors when submitting the application in cluster
+          | mode.
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala]: and stored in a `jars` subdirectory in each process's CWD

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesSparkRestServer.scala [hunk 4]
```diff
@@ -149,37 +150,42 @@ private[spark] class KubernetesSparkRestServer(
...
                 val resolvedJars = writtenJars ++ originalJars ++ Array(appResourcePath)
                 val sparkJars = new File(sparkHome, "jars").listFiles().map(_.getAbsolutePath)
+                val driverExtraClasspath = sparkProperties
+                  .get("spark.driver.extraClassPath")
+                  .map(_.split(","))
+                  .getOrElse(Array.empty[String])
                 val driverClasspath = driverExtraClasspath ++
                   resolvedJars ++
-                  sparkJars ++
-                  Array(appResourcePath)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesSparkRestServer.scala:178]: why is this line removed?  should it not have been there before?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesSparkRestServer.scala [hunk 5]
```diff
@@ -149,37 +150,42 @@ private[spark] class KubernetesSparkRestServer(
...
-                resolvedSparkProperties ++= sparkProperties
+                  sparkJars
                 resolvedSparkProperties("spark.jars") = resolvedJars.mkString(",")
 
+                // Resolve spark.files
+                val originalFiles = sparkProperties.get("spark.files")
+                  .map(_.split(","))
+                  .getOrElse(Array.empty[String])
+                val resolvedFiles = originalFiles ++ writtenFiles
+                resolvedSparkProperties("spark.files") = resolvedFiles.mkString
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesSparkRestServer.scala]: does this need `(",")` as a parameter to the `mkString` ?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/kubernetes/integrationtest/KubernetesSuite.scala [hunk 6]
```diff
@@ -334,4 +340,73 @@ private[spark] class KubernetesSuite extends SparkFunSuite with BeforeAndAfter {
...
+      "--deploy-mode", "cluster",
+      "--kubernetes-namespace", NAMESPACE,
+      "--name", "spark-file-existence-test",
+      "--executor-memory", "512m",
+      "--executor-cores", "1",
+      "--num-executors", "1",
+      "--upload-jars", HELPER_JAR,
+      "--upload-files", TEST_EXISTENCE_FILE.getAbsolutePath,
+      "--class", FILE_EXISTENCE_MAIN_CLASS,
+      "--conf", "spark.ui.enabled=true",
```
Comment (ash211) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/kubernetes/integrationtest/KubernetesSuite.scala]: if you don't need the ui, disabling saves a decent amount of startup time

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala [hunk 7]
```diff
@@ -511,18 +513,40 @@ private[spark] class Client(
...
       secret = secretBase64String,
       sparkProperties = sparkConf.getAll.toMap,
-      uploadedJarsBase64Contents = uploadJarsBase64Contents)
+      uploadedJarsBase64Contents = uploadJarsBase64Contents,
+      uploadedFilesBase64Contents = uploadFilesBase64Contents)
+  }
+
+  // Because uploaded files should be added to the working directory of the driver, they
+  // need to not have duplicate file names. They are added to the working directory so the
+  // user can reliably locate them in their application.
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: note that this is the same pattern as YARN?  I believe it is

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala [hunk 8]
```diff
@@ -511,18 +513,40 @@ private[spark] class Client(
...
+  private def validateNoDuplicateUploadFileNames(): Unit = {
+    uploadedFiles.foreach { unsplitPaths =>
+      val splitPaths = unsplitPaths.split(",")
+      val allPathsByFileName = splitPaths.groupBy(new File(_).getName)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: if you're not using `splitPaths` and `allPathsByFileName` except on the way to creating `pathsWithDuplicateNames` I think it makes sense to not give them identifiers, and instead just chain straight through

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala [hunk 9]
```diff
@@ -61,7 +61,9 @@ private[spark] class Client(
   private val sslSecretsDirectory = s"$DRIVER_CONTAINER_SECRETS_BASE_DIR/$kubernetesAppId-ssl"
   private val sslSecretsName = s"$SUBMISSION_SSL_SECRETS_PREFIX-$kubernetesAppId"
   private val driverDockerImage = sparkConf.get(DRIVER_DOCKER_IMAGE)
-  private val uploadedJars = sparkConf.get(KUBERNETES_DRIVER_UPLOAD_JARS)
+  private val uploadedJars = sparkConf.get(KUBERNETES_DRIVER_UPLOAD_JARS).filter(_.nonEmpty)
+  private val uploadedFiles = sparkConf.get(KUBERNETES_DRIVER_UPLOAD_FILES).filter(_.nonEmpty)
+  validateNoDuplicateUploadFileNames()
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala]: this seems a little weird -- should we have that method take an input and return an output instead of relying on the state of the instance vars being set already?

**Review by ash211 — COMMENTED** [review]
Looks good!  Happy to see the test verifying that it's placed on the resulting driver pod too.

Some minor nitpicky stuff but I saw nothing major on this diff

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesSparkRestServer.scala (hunk 4 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesSparkRestServer.scala:178]: `sparkJars` shouldn't be removed - that was a bad merge. `appResourcePath`, however, is already added above.

Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesSparkRestServer.scala:178]: Actually never mind, misread the diff. `appResourcePath` is not necessary here.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala (hunk 3 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala]: The jars are actually stored in temp space, which makes it basically impossible to find by the user.

**Comment by mccheah** [comment]
Documentation is added, addressed @aash's comments

**Review by ash211 — APPROVED** [review]
This looks great to me!  @iyanuobidele / @foxish any last comments before merging?  I'd like to get this in mid-afternoon if possible.

**Comment by foxish** [comment]
LGTM

**Comment by iyanu-obidele** [comment]
Looks good to me!

**Comment by iyanuobidele** [comment]
LGTM

**Comment by ash211** [comment]
Cheers, merged!
